staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/raw.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cbor
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var sharedTranscoders transcoders
+
+var rawTypeTranscodeFuncs = map[reflect.Type]func(reflect.Value) error{
+	reflect.TypeFor[runtime.RawExtension](): func(rv reflect.Value) error {
+		if !rv.CanAddr() {
+			return nil
+		}
+		re := rv.Addr().Interface().(*runtime.RawExtension)
+		if re.Raw == nil {
+			// When Raw is nil it encodes to null. Don't change nil Raw values during
+			// transcoding, they would have unmarshalled from JSON as nil too.
+			return nil
+		}
+		j, err := re.MarshalJSON()
```

Comment:
So we're navigating every unmarshalled cbor concrete type, using reflection, to cover the 14 known cases of RawExtension?  Just how expensive is this server-side?

On the client-side, this is the behavior we were talking about changing defaults on, correct?

Reply:
server-side, could we fix this in a rest.Strategy (BeforeCreate/BeforeUpdate) and an integration test that attempts to write CBOR to every RawExtension and expects to have json in etcd?  How much of a performance boost do we get?

Reply:
Summarizing the out-of-band conversation on this here...

Types are traversed using reflection only the first time they're seen, in order to build a function for transcoding values of that type. The value transcode function is cached, and only traverses parts of a value that can hold a RawExtension. If it can be determined statically that a type can never hold a RawExtension (most types), its value transcode function is a no-op.

The traversal cost is only significant for types that are composed of many interface types, since the dynamic type of any single interface value could be almost anything.

Kube API types whose value transcode function is _not_ a no-op are:

```
v1.AdmissionRequest
v1.AdmissionReview
v1alpha2.AllocationResult
v1alpha2.DriverAllocationResult
v1alpha2.DriverRequests
v1alpha2.ResourceClaim
v1alpha2.ResourceClaimList
v1alpha2.ResourceClaimParameters
v1alpha2.ResourceClaimParametersList
v1alpha2.ResourceClaimStatus
v1alpha2.ResourceClassParameters
v1alpha2.ResourceClassParametersList
v1alpha2.ResourceHandle
v1alpha2.ResourceRequest
v1alpha2.StructuredResourceHandle
v1alpha2.VendorParameters
v1beta1.AdmissionRequest
v1beta1.AdmissionReview
v1beta1.ControllerRevision
v1beta1.ControllerRevisionList
v1beta2.ControllerRevision
v1beta2.ControllerRevisionList
v1.ControllerRevision
v1.ControllerRevisionList
v1.List
```